Update the schema for the REST API specification

karmi opened the pull request:
The purpose of this patch is to improve the schema of the [Elasticsearch REST API specification](https://github.com/elastic/elasticsearch/tree/master/rest-api-spec/src/main/resources/rest-api-spec/api).

It tries to solve a number of outstanding issues with the current specification, and to make the schema more open for future extension. It requires the consumers — eg. the client code generators — to update their code, but the long-term benefits should be worth it.

A brief overview of the changes:

* It enhances the support for deprecated paths, introduced in https://github.com/elastic/elasticsearch/pull/39063
* It changes the specification of URL paths: instead of a list of strings, it uses a list of objects, to encode eg. the `method` and `path` combination.
* It adds a sharp and concise description to each API endpoint, to be used eg. in generated code annotations.

A snippet of the new schema, copied from the "Create" API, looks like this:

```json
{
"url":{
  "paths":[
    {
      "path":"/{index}/_create/{id}",
      "method":"PUT",
      "parts":{
        "id":{
          "type":"string",
          "required":true,
          "description":"Document ID"
        },
        "index":{
          "type":"string",
          "required":true,
          "description":"The name of the index"
        }
      }
    },
    {
      "path":"/{index}/{type}/{id}/_create",
      "method":"PUT",
      "parts":{
        "id":{
          "type":"string",
          "required":true,
          "description":"Document ID"
        },
        "index":{
          "type":"string",
          "required":true,
          "description":"The name of the index"
        },
        "type":{
          "type":"string",
          "description":"The type of the document",
          "deprecated":true
        }
      },
      "deprecated":{
        "version":"7.0.0",
        "description":"Specifying types in urls has been deprecated"
      }
    }
  ]
}}
```

The URL paths are no longer defined as a list of strings, with disconnected `parts` and `method` elements outside the definition. Instead, every path now lists the HTTP method to use, and the dynamic parts it contains. It also defines whether this path has been deprecated, instead of using an extra key, `deprecated_paths`, as before. For the cost of a little duplication, each dynamic part can now be marked as deprecated, directly in the path definition.

The dubious top-level `path` element has been removed: the original idea was to consider it as the "canonical" path, but in practice, the specification consumers never had a good use for it, and the ambiguity between `path` and `paths` created a lot of confusion.

The `params` and `body` elements have been left intact, as well as the recently introduced `deprecated` element.

Another breaking change is that the `documentation` element has been transformed to an object. Apart from the existing `url` key, it adds a `description` key, which contains a concise description of the API endpoint. These descriptions are used in the client code generators, and they have to be written carefully — they are displayed prominently on the documentation websites for various languages, for example on `godoc.org`:

![api-description-godoc](https://user-images.githubusercontent.com/4790/58172205-2f2f1a00-7c98-11e9-8426-485e6eaef656.png)

To allow the code generators work with a freeform text like this, the description has to be a full sentence, and has to start with a verb. Multiple lines are allowed.

One outstanding question is whether we should drop support for multiple HTTP methods in the endpoint definition, as this patch does. After some thinking and conversations, I think it's the right thing to do: a mere list of methods doesn't contain any information if there's any difference (for example in case of the "Create" API), and in some cases (the "Search" API), the listed methods are included as "optional", trying to encode certain leniency in the Elasticsearch API. This is probably confusing to specification readers, and doesn't have much value in the end.

To make the updates, I've created a Ruby script to load the old JSON files ([json_spec_update.rb.txt](https://github.com/elastic/elasticsearch/files/3216393/json_spec_update.rb.txt)), and generate the files with a new schema. I'm using another Ruby script ([json_spec_retouch.rb.txt](https://github.com/elastic/elasticsearch/files/3216395/json_spec_retouch.rb.txt)) to "retouch" the new files, when needed.

Related:

* https://github.com/elastic/elasticsearch/issues/35262
* https://github.com/elastic/elasticsearch/pull/39063
* https://github.com/elastic/elasticsearch/pull/41439
* https://github.com/elastic/elasticsearch/pull/41444

/cc @elastic/es-clients

javanna commented:
I had a look and I think the changes are a good improvement of our spec. For instance, for years it was hard to associate paths with their corresponding parts. I think though that we should keep, at least for now, multiple methods. Take bulk for instance (but this applies to many API), it supports POST and PUT for all of its paths. I would not want to see only POST in the spec unless the corresponding code supports only POST. Or rather, if we do want to make such change, I would like to make it separately from this refactoring and give it more visibility, have a separate discussions about it and so on.

If you adapt that I can take a stab at updating the java code in the yaml test runner so we get a green build and we can move this forward.

karmi commented:
I've re-introduced the multiple methods for endpoint in c6193e1.

elasticmachine commented:
Pinging @elastic/es-core-infra

codebrain commented:
This is a move in a good direction for the clients team.
❤️

javanna commented on rest-api-spec/src/main/resources/rest-api-spec/api/create.json:
```diff
@@ -1,69 +1,104 @@
+          "parts":{
+            "id":{
+              "type":"string",
+              "required":true,
```
I think that we can get rid of the required flag here, now that parts are listed under their own path. All parts are then required in the path that they are part of.

karmi replied:
Yes, good good catch! The `required` property is indeed redunant in the new format, since effectively _all_ the dynamic parts in a specific path are required.

javanna commented:
heya I have adapted the yaml test runner and merged master in. Also left a comment on removing the `required` flag from path parts. The path resolution logic is much more straight-forward now, which demonstrates that this is a good change. Let's see if tests are green.

javanna commented:
Tests seem to be red because there are more spec files that need to be converted to the new format. I think that besides `./rest-api-spec/src/main/resources/rest-api-spec/api/`, spec are at least also under `./x-pack/plugin/src/test/resources/rest-api-spec/api/`, and there's one here too: `./plugins/examples/rest-handler/src/test/resources/rest-api-spec/api/cat.example.json`

karmi committed:
Update the REST API specification

This patch updates the REST API spefication in JSON files to better encode deprecated entities,
to improve specification of URL paths, and to open up the schema for future extensions.

Notably, it changes the `paths` from a list of strings to a list of objects, where each
particular object encodes all the information for this particular path: the `parts` and the `methods`.

Among the benefits of this approach is eg. encoding the difference between using the `PUT` and `POST`
methods in the Index API, to either use a specific document ID, or let Elasticsearch generate one.

Also `documentation` becomes an object that supports an `url` and also a `description` which is a
new field.

karmi commented:
I've re-generated the JSON files for Core APIs from current master (picking up the new `stability` field), generated the JSON files for X-Pack APIs, and cherry-picked the changes to the Java test runner.

karmi commented:
Jenkins, retest this.

karmi commented:
Jenkins, retest this please.

javanna committed:
Adapt YAML runner to new REST API specification format

The logic for choosing the path to use when running tests has been
simplified, as a consequence of the path parts being listed under each
path in the spec. The special case for create and index has been removed.

Also the parsing code has been hardened so that errors are thrown earlier
when the structure of the spec differs from what expected, and their
error messages should be more helpful.

javanna commented:
I fixed a couple of test issues I found:
- the index API needed to be adapted as endpoints without it don't support PUT
- the indices.close new API needed to be converted
- one failing unit test due to the addition of the "stability" field in the spec

and rebased the branch on current master, which had a lot of conflicts due to #44487 .

javanna commented:
retest this please

elasticcla commented:
Hi @karmi, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in your Git [commit](https://github.com/elastic/elasticsearch/pull/42346.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

javanna commented:
run elasticsearch-ci/bwc

Mpdreamz commented on rest-api-spec/src/main/resources/rest-api-spec/api/index.json:
```diff
@@ -10,6 +10,7 @@
         {
           "path":"/{index}/_doc/{id}",
           "methods":[
+            "POST",
```
The rule here was and should remain the first `HTTP method` in the array should be the preferred one, so i think `PUT` should come before `POST` here.

javanna replied:
aren't these methods left in the same order as they were before the change?

Mpdreamz replied:
`index.json` always needed special handling in our generators because methods was for the whole API not per endpoint. 

Being able to put the preferred HTTP method for `{index}/_doc/{id}` is one of the benefits of the new format :smiley:

javanna replied:
ok lets keep this as a follow-up please. I just got a green build on this huge change after merging, re-merging, running and re-running tests :)

Mpdreamz replied:
Cool cool, will submit it after we get this merged :+1:

javanna replied:
thanks!

javanna reviewed (approved):
I have reviewed the spec changes, and adapted the yaml spec parser and the test runner. The stuff that I did not write myself LGTM :) was good that many little things were found in the process which allowed to harden tests and parsing code.

cjcenizal commented:
Thank you for the clear PR description @karmi and for the broadcast email @Mpdreamz! They were both immensely helpful for the ES UI team.

I see a number of breaking changes here which impact the ES UI team's maintenance workflow for Kibana Console (addressed by https://github.com/elastic/kibana/pull/43427). This isn't terribly disruptive for us, but could we please add the `>breaking` label to this PR? And in the future would it be possible to add the `>breaking` label to any PRs that change an interface published by this repo, including the REST API spec? Thanks!
